internal/agent/device/engine.go

```diff
@@ -2,93 +2,101 @@ package device
 
 import (
 	"context"
+	"sync"
 	"time"
 
 	"github.com/flightctl/flightctl/internal/util"
 )
 
+type Unit struct {
+	Interval time.Duration
+	Fn       func(context.Context)
+	lastRun  time.Time
+}
+
+// NewUnit creates a new unit with the given interval and function.
+func NewUnit(interval time.Duration, fn func(context.Context)) *Unit {
+	return &Unit{
+		Interval: interval,
+		Fn:       fn,
+	}
+}
+
+func (u *Unit) ready(now time.Time) bool {
+	return u.Interval > 0 && now.Sub(u.lastRun) >= u.Interval
+}
+
+func (u *Unit) run(ctx context.Context, now time.Time) {
+	u.Fn(ctx)
+	u.lastRun = now
+}
+
 type Engine struct {
-	fetchSpecInterval  util.Duration
-	fetchSpecFn        func(context.Context)
-	pushStatusInterval util.Duration
-	pushStatusFn       func(context.Context)
+	units []*Unit
```

Comment:
engine logic change does not seem necessary and could lead to future confusion.

Reply:
If it seems confusing it can be reverted.  For the notifier a similar logic needs to be created.